lib/zookeeper-proxy.js

```diff
@@ -37,26 +37,33 @@ module.exports = CoreObject.extend({
     var client = this.client;
     var _createPromisesHash = this._createPromisesHash;
 
-    return client.exists(key).then(function(res) {
-      if (res.stat) {
-        return;
-      }
+    return this.connect()
+      .then(client.exists.bind(client, key))
+      .then(function(res) {
+        if (res.stat) {
+          return;
+        }
 
-      // If it doesn't exist, memoize the actual creation
-      // so that this request to create can only happen once.
-      if (!(key in _createPromisesHash)) {
-        _createPromisesHash[key] = client.create(key);
-      }
+        // If it doesn't exist, memoize the actual creation
+        // so that this request to create can only happen once.
+        if (!(key in _createPromisesHash)) {
+          _createPromisesHash[key] = client.create(key);
+        }
 
-      return _createPromisesHash[key];
-    });
+        return _createPromisesHash[key];
+      });
   },
 
-  connect: function(connectionTimeout) {
+  connect: function() {
+    var connectionTimeout = this.options.connectionTimeout || 10000;
     var client = this.client;
     var timeout;
 
-    return new Promise(function(resolve, reject) {
+    if (this.connection) {
```

Comment:
I think that's not a bad idea. I'm going to skip it for the time being as this `zookeeper-proxy` is really a private class and nobody who uses this addon should ever need to access any of these things directly. (But, if they do, it's certainly worth revisiting!)